src/directory/s3_directory.rs

```diff
@@ -0,0 +1,285 @@
+use common::make_io_err;
+use directory::Directory;
+use directory::error::{IOError, OpenWriteError, OpenReadError, DeleteError, OpenDirectoryError};
+use directory::ReadOnlySource;
+use directory::shared_vec_slice::SharedVecSlice;
+use directory::WritePtr;
+use std::collections::HashMap;
+use std::convert::From;
+use std::default::Default;
+use std::error::Error;
+use std::fmt;
+use std::io;
+use std::path::{Path, PathBuf};
+use std::result;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::sync::RwLock;
+use rusoto_core::{DefaultCredentialsProvider, Region, default_tls_client};
+use rusoto_s3::{S3, S3Client, HeadBucketRequest, GetObjectRequest};
+
+fn get_client(region: Region) -> Result<Box<S3>, Box<Error>> {
+    // TODO: handle missing creds
+    let client = default_tls_client()?;
+    let provider = DefaultCredentialsProvider::new()?;
+
+    Ok(Box::new(S3Client::new(client, provider, region)))
+}
+
+#[derive(Clone)]
+struct InnerDirectory {
+    bucket: String,
+    cache: Arc<RwLock<HashMap<PathBuf, Arc<Vec<u8>>>>>,
```

Comment:
I think you can make the directory even nicer if instead of using this cache in memory, you wrapped another directory.

So to get the same functionality at this one, you would pass it the RAMDirectory, but if you wanted your cache to be persistant, you could use the MmapDirectory.